File: cast-extractor/src/main/java/com/libreshockwave/tools/preview/ScriptPreview.java
package com.libreshockwave.tools.preview;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.chunks.ScriptChunk;
import com.libreshockwave.chunks.ScriptNamesChunk;
import com.libreshockwave.tools.format.InstructionFormatter;
import com.libreshockwave.tools.model.CastMemberInfo;
import com.libreshockwave.tools.scanning.MemberResolver;

import java.util.ArrayList;
import java.util.List;

/**
 * Generates script preview content with bytecode disassembly.
 */
public class ScriptPreview {

    /**
     * Formats script content for display.
     */
    public String format(DirectorFile dirFile, CastMemberInfo memberInfo) {
        ScriptChunk script = MemberResolver.findScriptForMember(dirFile, memberInfo.member());
        ScriptNamesChunk names = dirFile.getScriptNames();

        StringBuilder sb = new StringBuilder();
        sb.append("=== SCRIPT: ").append(memberInfo.name()).append(" ===\n\n");
        sb.append("Member ID: ").append(memberInfo.memberNum()).append("\n");

        if (script == null) {
            sb.append("\n[No bytecode found for this script member]\n");
        } else {
            sb.append("Script Type: ").append(MemberResolver.getScriptTypeName(script.scriptType())).append("\n");
            sb.append("Behavior Flags: 0x").append(Integer.toHexString(script.behaviorFlags())).append("\n\n");

            // Properties
            if (!script.properties().isEmpty()) {
                sb.append("--- PROPERTIES ---\n");
                for (ScriptChunk.PropertyEntry prop : script.properties()) {
                    String propName = names != null ? names.getName(prop.nameId()) : "#" + prop.nameId();
                    sb.append("  property ").append(propName).append("\n");
                }
                sb.append("\n");
            }

            // Globals
            if (!script.globals().isEmpty()) {
                sb.append("--- GLOBALS ---\n");
                for (ScriptChunk.GlobalEntry global : script.globals()) {
                    String globalName = names != null ? names.getName(global.nameId()) : "#" + global.nameId();
                    sb.append("  global ").append(globalName).append("\n");
                }
                sb.append("\n");
            }

            // Handlers
            sb.append("--- HANDLERS (").append(script.handlers().size()).append(") ---\n\n");
            for (ScriptChunk.Handler handler : script.handlers()) {
                formatHandler(sb, handler, script, names);
            }

            // Literals
            if (!script.literals().isEmpty()) {
                sb.append("--- LITERALS (").append(script.literals().size()).append(") ---\n");
                int idx = 0;
                for (ScriptChunk.LiteralEntry lit : script.literals()) {
                    String typeStr = switch (lit.type()) {
                        case 1 -> "string";
                        case 4 -> "int";
                        case 9 -> "float";
                        default -> "type" + lit.type();
                    };
                    String valueStr = lit.value() instanceof String ?
                            "\"" + lit.value() + "\"" : String.valueOf(lit.value());
                    sb.append("  [").append(idx++).append("] ").append(typeStr).append(": ").append(valueStr).append("\n");
                }
            }
        }

        return sb.toString();
    }

    private void formatHandler(StringBuilder sb, ScriptChunk.Handler handler,
                               ScriptChunk script, ScriptNamesChunk names) {
        String handlerName = names != null ? names.getName(handler.nameId()) : "#" + handler.nameId();

        // Build argument list
        List<String> argNames = new ArrayList<>();
        for (int argId : handler.argNameIds()) {
            argNames.add(names != null ? names.getName(argId) : "#" + argId);
        }
        String argsStr = String.join(", ", argNames);

        sb.append("on ").append(handlerName);
        if (!argsStr.isEmpty()) {
            sb.append(" ").append(argsStr);
        }
        sb.append("\n");

        // Local variables
        if (!handler.localNameIds().isEmpty()) {
            List<String> localNames = new ArrayList<>();
            for (int localId : handler.localNameIds()) {
                localNames.add(names != null ? names.getName(localId) : "#" + localId);
            }
            sb.append("  -- locals: ").append(String.join(", ", localNames)).append("\n");
        }

        // Bytecode instructions
        sb.append("  -- bytecode (").append(handler.bytecodeLength()).append(" bytes, ")
                .append(handler.instructions().size()).append(" instructions):\n");

        for (ScriptChunk.Handler.Instruction instr : handler.instructions()) {
            sb.append("    ").append(InstructionFormatter.format(instr, script, names)).append("\n");
        }

        sb.append("end\n\n");
    }
}

File: cpp/src/editor/preview/ScriptPreview.cpp
#include "libreshockwave/editor/preview/ScriptPreview.hpp"

#include <iomanip>
#include <sstream>
#include <string>
#include <vector>

#include "libreshockwave/chunks/ScriptNamesChunk.hpp"
#include "libreshockwave/editor/format/InstructionFormatter.hpp"
#include "libreshockwave/editor/preview/PreviewFormatUtils.hpp"
#include "libreshockwave/editor/scanning/MemberResolver.hpp"
#include "libreshockwave/format/ScriptFormatUtils.hpp"

namespace libreshockwave::editor::preview {
namespace {

std::string resolveName(const chunks::ScriptNamesChunk* names, int nameId) {
    if (names != nullptr) {
        return names->getName(nameId);
    }
    return "#" + std::to_string(nameId);
}

std::string join(const std::vector<std::string>& values, const std::string& separator) {
    std::string result;
    for (const auto& value : values) {
        if (!result.empty()) {
            result += separator;
        }
        result += value;
    }
    return result;
}

std::string hexString(int value) {
    std::ostringstream out;
    out << std::hex << value;
    return out.str();
}

} // namespace

std::string ScriptPreview::format(DirectorFile& dirFile, const model::CastMemberInfo& memberInfo) const {
    const auto script = scanning::MemberResolver::findScriptForMember(dirFile, memberInfo.member);
    const auto names = dirFile.scriptNames();

    std::string out;
    PreviewFormatUtils::appendMemberHeader(out, "SCRIPT", memberInfo, false);

    if (script == nullptr) {
        out += "\n[No bytecode found for this script member]\n";
        return out;
    }

    out += "Script Type: " + ::libreshockwave::format::getScriptTypeName(script->scriptType()) + "\n";
    out += "Behavior Flags: 0x" + hexString(script->behaviorFlags()) + "\n\n";

    if (!script->properties().empty()) {
        out += "--- PROPERTIES ---\n";
        for (const auto& property : script->properties()) {
            out += "  property " + resolveName(names.get(), property.nameId) + "\n";
        }
        out += "\n";
    }

    if (!script->globals().empty()) {
        out += "--- GLOBALS ---\n";
        for (const auto& global : script->globals()) {
            out += "  global " + resolveName(names.get(), global.nameId) + "\n";
        }
        out += "\n";
    }

    out += "--- HANDLERS (" + std::to_string(script->handlers().size()) + ") ---\n\n";
    for (const auto& handler : script->handlers()) {
        formatHandler(out, handler, *script, names.get());
    }

    if (!script->literals().empty()) {
        out += "--- LITERALS (" + std::to_string(script->literals().size()) + ") ---\n";
        int index = 0;
        for (const auto& literal : script->literals()) {
            out += "  [" + std::to_string(index++) + "] " + ::libreshockwave::format::formatLiteral(literal) + "\n";
        }
    }

    return out;
}

void ScriptPreview::formatHandler(std::string& out,
                                  const chunks::ScriptChunk::Handler& handler,
                                  const chunks::ScriptChunk& script,
                                  const chunks::ScriptNamesChunk* names) const {
    std::vector<std::string> argNames;
    argNames.reserve(handler.argNameIds.size());
    for (int nameId : handler.argNameIds) {
        argNames.push_back(resolveName(names, nameId));
    }

    out += "on " + resolveName(names, handler.nameId);
    const auto args = join(argNames, ", ");
    if (!args.empty()) {
        out += " " + args;
    }
    out += "\n";

    if (!handler.localNameIds.empty()) {
        std::vector<std::string> localNames;
        localNames.reserve(handler.localNameIds.size());
        for (int nameId : handler.localNameIds) {
            localNames.push_back(resolveName(names, nameId));
        }
        out += "  -- locals: " + join(localNames, ", ") + "\n";
    }

    out += "  -- bytecode (" + std::to_string(handler.bytecodeLength) + " bytes, " +
           std::to_string(handler.instructions.size()) + " instructions):\n";
    for (const auto& instruction : handler.instructions) {
        out += "    " + ::libreshockwave::editor::format::InstructionFormatter::format(instruction, script, names) + "\n";
    }
    out += "end\n\n";
}

} // namespace libreshockwave::editor::preview
